### ISSUE TITLE
Unimplemented NumPy core functions

### ISSUE THREAD

**Issue by alexbw** [google/jax#70]
Remaining functions to be implemented:

- [x] np.argpartition (tricky because XLA does not provide any partial sort functionality) 
- [x] np.partition (tricky because XLA does not provide any partial sort functionality)

The list above was made by inspecting `jnp._NOT_IMPLEMENTED` and excluding deprecated functions (such as `np.alen`, `np.ipmt`, etc.), functions not relevant to JAX (such as `np.setbufsize`, `np.ascontiguousarray`, etc), and functions that modify buffers in-place (`np.put`, `np.place`, etc.):

Bugs for high-level categories:
 - [x] np.linalg: https://github.com/google/jax/issues/1999
 - [x] np.fft: https://github.com/google/jax/issues/1877
 - [x] searching: https://github.com/google/jax/issues/2080
 - [x] sorting: https://github.com/google/jax/issues/2079
 - [x] sets: https://github.com/google/jax/issues/2078
 - [x] NaN functions: https://github.com/google/jax/issues/2077
 - [x] Index tricks: https://github.com/google/jax/issues/5850

**Comment by mattjj** [comment]
The list needs a bit of refinement, since some things were already implemented and other things are out of scope, like to do with loading/saving/string manipulation. I deleted a few of the out-of-scope ones, and marked as done some things that were already included and some things that are coming in #96.

**Comment by rainwoodman** [comment]
Is there a guide on how to write customized primitive functions (or is that even possible?). I am looking for the equivalent to autograd's `defvjp`.

A quick glance into the code suggests everything is delegated to XLA -- if an operator is nontrivial then it has to be first implemented in XLA?

**Comment by kovasb** [comment]
Looking at contributing a function or two from this list. 

What is the approach to functions with nonstatically-sized return values? For example, setxor1d. 

My understanding is XLA needs to compile for each specific shape. Is there a pattern for avoiding worse-possible-case behavior, or should functions like setxor1d be removed from this list?

**Comment by souravsingh** [comment]
@mattjj I am looking to take a stab at the issue. How do I start?

**Comment by hawkinsp** [comment]
@souravsingh Awesome! In general, lots of these are fairly easy to do. Pick one you like off the list, and take a look at previous PRs implementing numpy ops. For example, I just sent out https://github.com/google/jax/pull/298 adding np.cumprod/cumsum support. Usually, it's just a question of implementing a numpy op in terms of the primitives in the `lax.py` library.

**Comment by navneet-nmk** [comment]
@mattjj I am looking to implement np.unique. Is this something which is a priority at the moment?

**Comment by mattjj** [comment]
Sounds like a great idea! The core team isn't working on any numpy operations at the moment (we're mostly focused on improving performance in the XLA runtime, parallel computation, and Cloud TPU support) so that's a great place to make contributions.

`np.unique` is tricky though because AIUI the shape of the output depends on the values of the input. Since XLA can only express computations with static shapes (meaning independent of the input values), we'd have to re-compile the XLA computation for every new value of the input, and moreover it might be tricky to express in terms of XLA HLO (i.e. in terms of `lax` library calls).

@hawkinsp any thoughts on implementing `np.unique`?

**Comment by navneet-nmk** [comment]
@mattjj since this would be my first hand at this, I should probably look at numpy functions that are less tricky compared to unique?

Any suggestions would be really helpful.

**Comment by navneet-nmk** [comment]
@mattjj Apologies for the constant posts. I have decided to implement np.diff which seems fairly easier. I was just wondering whether there is a guide to developing the testcases in the lax_numpy_test.py file. Thanks!

**Comment by murphyk** [comment]
+1 to adding np.cov

**Comment by murphyk** [comment]
+1 to np.percentile

**Comment by mattjj** [comment]
@murphyk np.cov coming in #983

**Comment by hawkinsp** [comment]
@murphyk `quantile` and `percentile` are now present, at least for `interpolation='linear'` and floating-point types.

**Comment by adarob** [comment]
fyi, np.rank is deprecated in favor of np.ndim

**Comment by jessebett** [comment]
Anyone interested in `np.convolve` I've commented on possibly using `lax.conv`. It would require reshaping and flipping the inputs. I've described it in #1561.

**Comment by michiboo** [comment]
@jessebett  I would like to try np.convolve if you are not working on it

**Comment by fedden** [comment]
I'd also like to get involved in this process of completing the remaining numpy ops. As a first pass, I am looking at implementing something really simple, to dip my toes in the water before tackling something more serious!

I will look at np.alen unless there are any issues with that :)

**Comment by fedden** [comment]
Created a PR for alen in #1924

**Comment by shoyer** [comment]
I took another pass at deleting irrelevant functions:
- anything currently deprecated by numpy
- anything that relies upon details of NumPy's internal array storage (strides, etc)
- anything related to `np.matrix`.
- anything that modifies arguments in place

I think most of the rest should be relevant, though some will be tricky (e.g., with output dependent shapes).

**Comment by StephenHogg** [comment]
Hi - I'm interested in getting `nanmean` done. Am I right in suggesting that this could be done by somehow piggypacking off `nancumsum`? If so, are there any tips that may be worth heeding in this regard?

**Comment by shoyer** [comment]
NumPy should be a pretty good model for most NaN functions, all of its NaN functions are written in pure Python.

**Comment by shoyer** [comment]
To make things a little easier to keep track of, I split some functions off into a handful of sub-issues:

- np.linalg: #1999
- np.fft: #1877
- searching: #2080
- sorting: #2079
- sets: #2078
- NaN functions: #2077

**Comment by oliverastrand** [comment]
Hi, should [np.asanyarray](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.asanyarray.html) be removed since it simply becomes [np.asarray](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.asarray.html) when there is no subclassing?

**Comment by ayush-1506** [comment]
Hi, quick question: I needed `np.fmax` and it looks like this is left to be implemented. Looking at the operators defined in lax.py, isn't `np.fmax` essentially going to be the same as `lax.max`? If so, should I open a PR for the same?

**Comment by jameskirkpatrick** [comment]
np.cast is not implemented in Jax. this is a dict of lambdas that allows casting to certain types.

**Comment by hawkinsp** [comment]
@jameskirkpatrick I agree that `np.cast` exists in my copy of NumPy, but it is not a documented API as far as I can tell. Are we sure it's an intentional, non-deprecated NumPy API? @shoyer any thoughts?

**Comment by bhushan-borole** [comment]
Hello @hawkinsp 
I am willing to take up `np.insert`.
Is someone else working on it?

**Comment by jakevdp** [comment]
I don't know of anyone working on these functions, although I think `min_scalar_type` cannot be made compatible with JIT because of its semantics. Please let us know if you have questions!

**Comment by jakevdp** [comment]
I updated the list to only include remaining functions that would be good candidates for implementation in JAX.

**Comment by avani17101** [comment]
np.poly, np.polyfit and np.poly1d have transitioned as mentioned in https://numpy.org/doc/stable/reference/routines.polynomials.html. 
Shall we transition it as well?

**Comment by jakevdp** [comment]
I'm not sure whether we'd want to try to replicate numpy's polynomial interface in JAX. The nice thing about `np.poly*` is that they basically just operate on arrays, which fits JAX's model more readily that manipulation of custom polynomial objects.

**Comment by ljjsalt** [comment]
Hi @jakevdp @hawkinsp!
I am interested in implementing ```np.polydiv```. Is someone else working on it?

**Comment by jakevdp** [comment]
There was a start on it in #7729, but that PR seems to have stalled. Feel free to work on it!

**Comment by Gairick52** [comment]
@alexbw Hello sir i want to contribute can you point me to some resources

**Comment by jakevdp** [comment]
Hi @Gairick52 - I removed the "good first issue" label because this issue is nearly complete, and the remaining TODOs are not trivial.

As far as contributing to JAX, the best approach I think would be to find something that overlaps with your own areas of expertise. What kinds of things do you use JAX for?

**Comment by alexbw** [comment]
https://youtu.be/xP8tFFJrtXU
